### PR TITLE
Save Image Link

### LINKED ISSUE #6: Save Image Link
#### As a user I want to save a link to an image.

_Assignees_: @HaominZhang, @XanderStrike

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

### PR DESCRIPTION
**Pull request by b00611005**
Closes #6 
Implement the functionality to display and save image link.

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -1,3 +1,25 @@
...
+      render 'new'
+    end
+  end
+
+  def show
+    @image = Image.find(params[:id])
+  end
+end
+
+private
```
Comment (HaominZhang) [inline app/controllers/images_controller.rb]: Minor but it's better to have a new line after private to make it more apparent in this class

File: db/migrate/20180221235822_create_images.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration[5.1]
+  def change
+    create_table :images do |t|
+      t.string :url, null: false
```
Comment (HaominZhang) [inline db/migrate/20180221235822_create_images.rb]: 👍

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
     # render plain: params[:image].inspect
     @image = Image.new(image_params)
 
     if @image.save
-      puts 'URL saved'
+      # logger.debug 'URL saved'
       redirect_to @image
     else
-      puts 'Invalid URL'
+      # logger.debug 'Invalid URL'
```
Comment (HaominZhang) [inline app/controllers/images_controller.rb]: I understand that we already have a way to tell the user if it is failed to save an image by using the `errors` attribute on the model, but I think it would be better if we could use flash message here and for success case as well so it is more obvious and user friendly. How do you think?

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -1,15 +1,17 @@
 class ImagesController < ApplicationController
-  def new; end
+  def new
+    @image = Image.new
+  end
 
   def create
     # render plain: params[:image].inspect
```
Comment (HaominZhang) [inline app/controllers/images_controller.rb]: If not really necessary, let's remove this comment

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -1,3 +1,25 @@
...
+      puts 'URL saved'
+      redirect_to @image
+    else
+      puts 'Invalid URL'
+      render 'new'
+    end
+  end
+
+  def show
+    @image = Image.find(params[:id])
```
Comment (HaominZhang) [inline app/controllers/images_controller.rb]: From programatic perspective, is it possible that we could access an image that doesn't exist? If yes, how do we want to handle it?

File: db/migrate/20180221235822_create_images.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration[5.1]
```
Comment (HaominZhang) [inline db/migrate/20180221235822_create_images.rb]: I think migration and schema could be in its own commit so if anything goes wrong with it, we can simply revert it without affecting any other code changes. How do you think?

File: test/models/image_test.rb [hunk 7]
```diff
@@ -1,7 +1,40 @@
 require 'test_helper'
```
Comment (HaominZhang) [inline test/models/image_test.rb]: I think the best practice for adding a test would be having it together with the original implementation instead of having a separate commit to add all tests. For example, unit test should go with the creation with the model and controller tests should go with the introduction of controller methods. Does it make sense to you?

File: test/models/image_test.rb [hunk 8]
```diff
@@ -1,7 +1,40 @@
...
+    assert_predicate @image, :valid?
+
+    @image = Image.new(url: 'https://mickie.gif')
+    assert_predicate @image, :valid?
+
+    @image = Image.new(url: 'https://mickie.png')
+    assert_predicate @image, :valid?
+  end
+
+  def test_invalid__url
```
Comment (HaominZhang) [inline test/models/image_test.rb]: How about adding assertions about blank url?

File: app/controllers/images_controller.rb (hunk 5 above)
Comment (b00611005) [reply app/controllers/images_controller.rb]: Yes, it's possible. The users could access an image that does not exist by directly changing the url with an nonexistent id. I replaced **find** with **find_by_id**, which return *nil* if it cannot find the image with that id. Then, check **@image** , if it's a **nil** then redirect users to the homepage.

File: test/models/image_test.rb (hunk 8 above)
Comment (b00611005) [reply test/models/image_test.rb]: Added some new tests for validating **blank url** and **url with only spaces**.

File: app/controllers/images_controller.rb (hunk 5 above)
Comment (HaominZhang) [reply app/controllers/images_controller.rb]: Another way to do it is keep the `find` method but `rescue` ActiveRecord::RecordNotFound exception. Just another thought :)

File: test/models/image_test.rb (hunk 7 above)
Comment (b00611005) [reply test/models/image_test.rb]: Totally makes sense. I'll do it next time.

File: db/migrate/20180221235822_create_images.rb (hunk 6 above)
Comment (b00611005) [reply db/migrate/20180221235822_create_images.rb]: Do you mean to make a commit just after `rails db:migrate` and related files created?

File: app/controllers/images_controller.rb (hunk 5 above)
Comment (b00611005) [reply app/controllers/images_controller.rb]: I found that rubocop asked me to `Use find_by instead of dynamic find_by_id.`, so I do the `rescue` one.

File: app/controllers/images_controller.rb (hunk 3 above)
Comment (b00611005) [reply app/controllers/images_controller.rb]: Yap, I agree that using flash message is a good idea and absolutely more user friendly than the `errors` attribute. I modified the codes to use flash message for both successful and unsuccessful cases.

File: app/controllers/images_controller.rb (hunk 1 above)
Comment (b00611005) [reply app/controllers/images_controller.rb]: Got you and fixed.

File: db/migrate/20180221235822_create_images.rb (hunk 2 above)
Comment (b00611005) [reply db/migrate/20180221235822_create_images.rb]: 👍

File: app/controllers/images_controller.rb (hunk 4 above)
Comment (b00611005) [reply app/controllers/images_controller.rb]: Fixed.
